src/server/pfs/server/driver.go

```diff
@@ -1716,43 +1720,39 @@ func (d *driver) dropCommit(ctx context.Context, txnCtx *txncontext.TransactionC
 	return nil
 }
 
+// fillNewBranches helps create the upstream branches on which a branch is provenant, if they don't exist.
 // TODO(provenance): consider removing this functionality
 func (d *driver) fillNewBranches(ctx context.Context, txnCtx *txncontext.TransactionContext, branch *pfs.Branch, provenance []*pfs.Branch) error {
+	branch.Repo.EnsureProject()
 	newRepoCommits := make(map[string]*pfsdb.CommitWithID)
 	for _, p := range provenance {
-		if branch.Repo.Project == nil {
-			branch.Repo.Project = &pfs.Project{Name: "default"}
-		}
-		branchInfo, err := pfsdb.GetBranchInfoByName(ctx, txnCtx.SqlTx, p.Repo.Project.Name, p.Repo.Name, p.Repo.Type, p.Name)
+		biWithID, err := pfsdb.GetBranchInfoWithID(ctx, txnCtx.SqlTx, p)
 		if err != nil && !pfsdb.IsNotFoundError(err) {
-			return errors.Wrap(err, "fill new branches")
+			return errors.Wrap(err, "get branch info")
+		}
+		if biWithID != nil {
+			// Skip because the branch already exists.
+			continue
 		}
 		var updateHead *pfsdb.CommitWithID
-		if branchInfo == nil {
-			branchInfo = &pfs.BranchInfo{}
-			branchInfo.Branch = p
-			var head *pfsdb.CommitWithID
-			var ok bool
-			if head, ok = newRepoCommits[pfsdb.RepoKey(branchInfo.Branch.Repo)]; !ok {
-				var err error
-				head, err = d.makeEmptyCommit(ctx, txnCtx, branchInfo.Branch, nil, nil)
-				if err != nil {
-					return err
-				}
-				updateHead = head
-				newRepoCommits[pfsdb.RepoKey(branchInfo.Branch.Repo)] = head
+		head, ok := newRepoCommits[p.Repo.Key()]
+		if !ok {
+			head, err = d.makeEmptyCommit(ctx, txnCtx, p, nil, nil)
+			if err != nil {
+				return errors.Wrap(err, "create empty commit")
 			}
-			branchInfo.Head = head.CommitInfo.Commit
+			updateHead = head
+			newRepoCommits[p.Repo.Key()] = head
 		}
+		branchInfo := &pfs.BranchInfo{Branch: p, Head: head.CommitInfo.Commit}
 		if _, err := pfsdb.UpsertBranch(ctx, txnCtx.SqlTx, branchInfo); err != nil {
-			return errors.Wrap(err, "fill new branches")
+			return errors.Wrap(err, "upsert branch")
```

Comment:
[nit] When wrapping the error, the wrap message should match the name of the function being called fillNewBranches(), because within UpsertBranch(), errors should also wrapped with a prefix 'upsert branch:'

Reply:
No, we wrap errors for two reasons:
 1) to add stacktrace
 2)  to add context to the function that received the error.

This way, among a slew of error handling within a function, it is clear exactly which line errored. For example, if you call `foo()` below and get back an error, it's hard to know if it's from `step1()` or `step2()`. However, if you add a `errors.Wrap(err, "step1")` then it's really easy to just grep the source code when you get the message.
```
func foo() error { 
    if err := step1(); err != nil { 
        return err
    }
    if err := step2(); err != nil {
        return err
    }
    return nil
}
```